### DISCUSSION TITLE
C++ wrapper plans

### DISCUSSION THREAD

**Discussion by dankamongmen** [dankamongmen/notcurses#1993] [Ideas]
@grendello did a great job early in Notcurses development with the initial C++ wrappers, and showed me some techniques -- I'm eternally appreciative. He came out of nowhere and dropped a complete set of high-quality wrappers, done pretty much entirely without my feedback, unrequested and unexpected.

If he ever returns to continue maintaining them, I'd be more than delighted to have him. He's been unreachable for a few months, though, in the way that people who don't have the bandwidth to deal with Yet Another Thing become unreachable. In no way am I anything other than thankful. With that said, the C++ wrappers are in need of some love.

Most critical in my mind is eliminating the `map<Plane>` that attempts to ensure a bijection between `ncplane`s and `ncpp::Plane`s. i understand what Marek was going for here, but i think it unsound. It leads to problems with destructors and other issues, and certainly complicates the wrappers. all it really seems to buy us (and here's where i'm maybe missing something) is that `Plane`s form a graph isomorphic to the `ncplane`s--i.e. you can compare two `Plane` pointers and they'll be the same if they refer to the same `ncplane`. i don't see that as terribly valuable, and certainly not worth the problems.

so i'd like to rip that out, which ought resolve several open bugs. i'd then like to extend the C++ wrappers out to cover new functionality and match recent deprecations.

open c++ bugs:
* #1983 
* #1702 
* #1474 
* #1312 
* #645

**Comment by WSLUser** [comment]
This seems a good spot to ask if and how allocations are done with the C++ wrappers. Do the wrappers create any allocations or does it all come from the original C code? The reason I ask is I randomly stumbled upon an MS project that boosts performance of allocations for C++ applications. https://github.com/microsoft/snmalloc It works on Windows, Linux, and BSD, even Dragonfly so seems like it would a perfect useful library to use if it's applicable here.

**Reply by dankamongmen** [reply]
i wouldn't have expected them to do any allocations, but as it turns out there is a Plane object corresponding to each `ncplane` see #1983. regardless, it's by no means allocation-heavy. really, nothing is -- all our allocations come in pools.

**Reply by WSLUser** [reply]
I figured it wouldn't really make sense here but wanted to be sure. I did bring it up for Contour and Windows Terminal, who most certainly do a ton of allocations for various things.
